File: zoopla-scraper/browser/python/zoopla.py
"""Zoopla scraper using the official Scrapeless Python SDK + Playwright over CDP.
function names + emitted field names match verbatim.

Two surfaces:
- `scrape_properties(urls)` — DOM-parse a Zoopla property page (no clean cache).
- `scrape_search(...)`      — DOM-parse cards + `__ZAD_TARGETING__` totals JSON.

Uses a GB residential proxy (Zoopla is geo-locked + Akamai-protected).
"""

from __future__ import annotations

import json
import os
from typing import Any, Literal

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "GB"
DEFAULT_SESSION_TTL = 180

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _new_browser(proxy_country: str = DEFAULT_PROXY_COUNTRY) -> str:
    client = _client()
    session = client.browser.create(
        ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
    )
    return session.browser_ws_endpoint

async def _fetch_html(url: str, ready_selector: str | None, *, auto_scroll: bool = False) -> str:
    ws = await _new_browser()
    async with async_playwright() as pw:
        browser = await pw.chromium.connect_over_cdp(ws)
        try:
            page = await browser.new_page()
            await page.goto(url, wait_until="domcontentloaded", timeout=45000)
            if ready_selector:
                try:
                    await page.wait_for_selector(ready_selector, timeout=15000)
                except Exception as e:
                    logger.warning("wait_for_selector failed: {}", e)
            if auto_scroll:
                # Trigger lazy-loaded sections before grabbing HTML.
                try:
                    await page.evaluate(
                        """async () => {
                            const sleep = ms => new Promise(r => setTimeout(r, ms));
                            const step = 600;
                            for (let y = 0; y < document.body.scrollHeight; y += step) {
                                window.scrollTo(0, y);
                                await sleep(150);
                            }
                            window.scrollTo(0, 0);
                        }"""
                    )
                    await page.wait_for_timeout(1000)
                except Exception as e:
                    logger.warning("auto-scroll failed: {}", e)
            return await page.content()
        finally:
            try:
                await browser.close()
            except Exception:
                pass

# ---------------------------------------------------------------------------
# Parsers — mirror the upstream reference verbatim
# ---------------------------------------------------------------------------

def parse_property(html: str) -> dict:
    """"""
    selector = Selector(text=html)
    url = selector.xpath("//meta[@property='og:url']/@content").get()
    price = selector.xpath("//p[contains(text(),'£')]/text()").get()
    receptions = selector.xpath("//p[contains(text(),'reception')]/text()").get()
    baths = selector.xpath("//p[contains(text(),'bath')]/text()").get()
    beds = selector.xpath("//p[contains(text(),'bed')]/text()").get()
    gmap_source = selector.xpath(
        "(//section[@aria-labelledby='local-area']//picture/source/@srcset)[last()]"
    ).get()
    coordinates = gmap_source.split("/static/")[1].split("/")[0] if gmap_source else None
    agent_path = selector.xpath("//section[@aria-label='Contact agent']//a/@href").get()

    info = []
    for i in selector.xpath("//section[@aria-labelledby='key-info']//li"):
        title = i.xpath(".//p/text()").get()
        value = i.xpath(".//div/p/text()").get()
        if value is not None:
            info.append({"title": title, "value": value})

    nearby = []
    for i in selector.xpath("//div[section[contains(@aria-label,'Travel')]]/section[3]//li/div"):
        distance = i.xpath(".//p[2]/text()").get()
        nearby.append(
            {
                "title": i.xpath(".//p[1]/text()").get(),
                "distance": float(distance.split(" ")[0]) if distance else None,
                "unit": distance.split(" ")[1] if distance else None,
            }
        )

    return {
        "id": int(url.split("details/")[-1].split("/")[0]) if url and "details/" in url else None,
        "url": url,
        "title": selector.xpath("//title/text()").get(),
        "address": selector.xpath("//address/text()").get(),
        "price": {
            "amount": int(price.replace("£", "").replace(",", "")) if price else None,
            "currency": "£",
        },
        "gallery": selector.xpath(
            "//li[contains(@data-key,'gallery')]/picture/source[last()]/@srcset"
        ).getall(),
        "epcRating": selector.xpath("//p[contains(text(),'EPC')]/text()").get(),
        "floorArea": selector.xpath("//p[contains(text(),'ft')]/text()").get(),
        "numOfReceptions": int(receptions.split(" ")[0]) if receptions else None,
        "numOfBathrooms": int(baths.split(" ")[0]) if baths else None,
        "numOfBedrooms": int(beds.split(" ")[0]) if beds else None,
        "propertyTags": selector.xpath("(//section/ul)[1]/li/p/text()").getall(),
        "propertyInfo": info,
        "propertyDescription": selector.xpath(
            "//section[@aria-labelledby='about']/ul/li/p/span/text()"
        ).getall(),
        "coordinates": {
            "googleMapeSource": gmap_source,
            "latitude": float(coordinates.split(",")[0]) if coordinates else None,
            "longitude": float(coordinates.split(",")[1]) if coordinates else None,
        },
        "nearby": nearby,
        "agent": {
            "name": selector.xpath("//section[@aria-label='Contact agent']//p/text()").get(),
            "logo": selector.xpath("//section[@aria-label='Contact agent']//img/@src").get(),
            "url": "https://www.zoopla.co.uk" + agent_path if agent_path else None,
        },
    }

def parse_search(html: str) -> dict:
    """Returns {search_data, total_pages}."""
    selector = Selector(text=html)
    data = []
    targeting = selector.xpath("//script[@id='__ZAD_TARGETING__']/text()").get()
    total_results = int(json.loads(targeting)["search_results_count"]) if targeting else 0
    boxes = selector.xpath("//div[@data-testid='regular-listings']/div")
    results_count = len(boxes) or 1
    total_pages = total_results // results_count

    for box in boxes:
        url = box.xpath(".//a/@href").get()
        if not url:
            continue
        price = box.xpath(".//p[contains(@class, 'priceText')]/text()").get()
        sq_ft = box.xpath(".//span[contains(text(),'sq ft')]/text()").get()
        sq_ft = int(sq_ft.split(" ")[0].strip("~,")) if sq_ft else None
        bathrooms = box.xpath(".//span[(contains(text(), 'bath'))]/text()").get()
        bedrooms = box.xpath(".//span[(contains(text(), 'bed'))]/text()").get()
        livingrooms = box.xpath(".//span[(contains(text(), 'reception'))]/text()").get()
        image = box.xpath(".//picture/source/@srcset").get()
        agency = box.xpath("//img[contains(@src,'agent')]/@alt").get()
        item = {
            "price": int(price.split(" ")[0].replace("£", "").replace(",", "")) if price else None,
            "priceCurrency": "Sterling pound £",
            "url": "https://www.zoopla.co.uk" + url.split("?")[0] if url else None,
            "image": image.split(":p")[0] if image else None,
            "address": box.xpath(".//address/text()").get(),
            "squareFt": sq_ft,
            "numBathrooms": int(bathrooms.split(" ")[0]) if bathrooms else None,
            "numBedrooms": int(bedrooms.split(" ")[0]) if bedrooms else None,
            "numLivingRoom": int(livingrooms.split(" ")[0]) if livingrooms else None,
            "description": box.xpath(".//a[address]/p/text()").get(),
            "justAdded": bool(box.xpath(".//div[text()='Just added']/text()").get()),
            "agency": agency,
        }
        data.append(item)
    return {"search_data": data, "total_pages": total_pages}

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference verbatim
# ---------------------------------------------------------------------------

async def scrape_properties(urls: list[str]) -> list[dict]:
    """Scrape Zoopla property listings."""
    out: list[dict] = []
    for url in urls:
        try:
            logger.info("scraping property page {}", url)
            html = await _fetch_html(
                url,
                ready_selector="section[aria-labelledby='local-area']",
                auto_scroll=True,
            )
            out.append(parse_property(html))
        except Exception as e:
            logger.error("error scraping {}: {}", url, e)
    return out

async def scrape_search(
    scrape_all_pages: bool,
    location_slug: str,
    max_scrape_pages: int = 10,
    query_type: Literal["for-sale", "to-rent"] = "for-sale",
) -> list[dict]:
    """Scrape Zoopla search pages for property listings."""
    first_url = f"https://www.zoopla.co.uk/{query_type}/property/{location_slug}"
    first_html = await _fetch_html(
        first_url,
        ready_selector="p[data-testid='total-results']",
        auto_scroll=True,
    )
    first_data = parse_search(first_html)
    search_data = list(first_data["search_data"])
    max_search_pages = first_data["total_pages"]
    total_pages_to_scrape = (
        max_search_pages if scrape_all_pages or max_scrape_pages >= max_search_pages
        else max_scrape_pages
    )
    logger.info("scraping {} more search pages", max(0, total_pages_to_scrape - 1))
    for p in range(2, total_pages_to_scrape + 1):
        try:
            html = await _fetch_html(
                f"{first_url}?pn={p}",
                ready_selector="div[data-testid='regular-listings']",
            )
            search_data.extend(parse_search(html)["search_data"])
        except Exception as e:
            logger.warning("failed page {}: {}", p, e)
    logger.info("scraped {} search listings from {}", len(search_data), first_url)
    return search_data

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    if isinstance(obj, dict):
        return {k: to_dict(v) for k, v in obj.items()}
    return obj

File: zoopla-scraper/browser/nodejs/zoopla.mjs
// Zoopla scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// function names + emitted field names match verbatim.
//
// Two surfaces:
// - `scrapeProperties(urls)`                                       — DOM-parse property page.
// - `scrapeSearch(scrapeAllPages, locationSlug, maxScrapePages, queryType)` — DOM-parse cards.
//
// Uses a GB residential proxy (Zoopla is geo-locked + Akamai-protected).

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "GB";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function newBrowser(proxyCountry = DEFAULT_PROXY_COUNTRY) {
  const { browserWSEndpoint } = await client().browser.create({
    proxyCountry,
    sessionTTL: DEFAULT_SESSION_TTL,
  });
  return browserWSEndpoint;
}

async function fetchHtml(url, readySelector, { autoScroll = false } = {}) {
  const ws = await newBrowser();
  let browser;
  try {
    browser = await puppeteer.connect({ browserWSEndpoint: ws });
    const page = await browser.newPage();
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
    if (readySelector) {
      try { await page.waitForSelector(readySelector, { timeout: 15000 }); } catch (_) {}
    }
    if (autoScroll) {
      try {
        await page.evaluate(async () => {
          const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
          const step = 600;
          for (let y = 0; y < document.body.scrollHeight; y += step) {
            window.scrollTo(0, y);
            await sleep(150);
          }
          window.scrollTo(0, 0);
        });
        await new Promise((r) => setTimeout(r, 1000));
      } catch (_) {}
    }
    return await page.content();
  } finally {
    try { await browser?.close(); } catch (_) {}
  }
}

const intOrNull = (s) => {
  if (!s) return null;
  const n = parseInt(String(s).split(" ")[0].replace(/[£,~]/g, ""), 10);
  return Number.isFinite(n) ? n : null;
};

// ---------------- parsers (mirror the upstream reference verbatim) ----------------

export function parseProperty(html) {
  const $ = cheerio.load(html);
  const url = $('meta[property="og:url"]').attr("content") ?? null;
  const price = $("p").filter((_, el) => /£/.test($(el).text())).first().text() || null;
  const receptions = $("p").filter((_, el) => /reception/.test($(el).text())).first().text() || null;
  const baths = $("p").filter((_, el) => /bath/.test($(el).text())).first().text() || null;
  const beds = $("p").filter((_, el) => /bed/.test($(el).text())).first().text() || null;
  const gmapSrcs = $("section[aria-labelledby='local-area'] picture source")
    .map((_, el) => $(el).attr("srcset"))
    .get();
  const gmapSource = gmapSrcs.length ? gmapSrcs[gmapSrcs.length - 1] : null;
  const coordinates = gmapSource && gmapSource.includes("/static/")
    ? gmapSource.split("/static/")[1].split("/")[0]
    : null;
  const agentPath = $("section[aria-label='Contact agent'] a").first().attr("href") ?? null;

  const info = [];
  $("section[aria-labelledby='key-info'] li").each((_, li) => {
    const $li = $(li);
    const title = $li.find("p").first().text() || null;
    const value = $li.find("div p").first().text() || null;
    if (value !== null && value !== "") info.push({ title, value });
  });

  const nearby = [];
  $("div:has(section[aria-label*='Travel']) section:nth-of-type(3) li div").each((_, div) => {
    const $div = $(div);
    const distance = $div.find("p").eq(1).text() || null;
    nearby.push({
      title: $div.find("p").first().text() || null,
      distance: distance ? parseFloat(distance.split(" ")[0]) : null,
      unit: distance ? distance.split(" ")[1] : null,
    });
  });

  const idFromUrl = url && url.includes("details/")
    ? parseInt(url.split("details/").pop().split("/")[0], 10)
    : null;

  return {
    id: Number.isFinite(idFromUrl) ? idFromUrl : null,
    url,
    title: $("title").first().text() || null,
    address: $("address").first().text() || null,
    price: {
      amount: price ? parseInt(price.replace(/£|,/g, ""), 10) : null,
      currency: "£",
    },
    gallery: $("li[data-key*='gallery'] picture source")
      .last()
      .toArray()
      .map((el) => $(el).attr("srcset"))
      .filter(Boolean),
    epcRating: $("p").filter((_, el) => /EPC/.test($(el).text())).first().text() || null,
    floorArea: $("p").filter((_, el) => /ft/.test($(el).text())).first().text() || null,
    numOfReceptions: intOrNull(receptions),
    numOfBathrooms: intOrNull(baths),
    numOfBedrooms: intOrNull(beds),
    propertyTags: $("section ul").first().find("li p").map((_, el) => $(el).text()).get(),
    propertyInfo: info,
    propertyDescription: $("section[aria-labelledby='about'] ul li p span")
      .map((_, el) => $(el).text())
      .get(),
    coordinates: {
      googleMapeSource: gmapSource,
      latitude: coordinates ? parseFloat(coordinates.split(",")[0]) : null,
      longitude: coordinates ? parseFloat(coordinates.split(",")[1]) : null,
    },
    nearby,
    agent: {
      name: $("section[aria-label='Contact agent'] p").first().text() || null,
      logo: $("section[aria-label='Contact agent'] img").first().attr("src") ?? null,
      url: agentPath ? "https://www.zoopla.co.uk" + agentPath : null,
    },
  };
}

export function parseSearch(html) {
  const $ = cheerio.load(html);
  const data = [];
  const targetingRaw = $("script#__ZAD_TARGETING__").first().text();
  let totalResults = 0;
  try {
    if (targetingRaw) totalResults = parseInt(JSON.parse(targetingRaw).search_results_count, 10) || 0;
  } catch (_) {}
  const boxes = $("div[data-testid='regular-listings'] > div");
  const resultsCount = boxes.length || 1;
  const totalPages = Math.floor(totalResults / resultsCount);

  boxes.each((_, box) => {
    const $box = $(box);
    const url = $box.find("a").first().attr("href");
    if (!url) return;
    const price = $box.find("p[class*='priceText']").first().text() || null;
    const sqFtText = $box.find("span").filter((_, el) => /sq ft/.test($(el).text())).first().text() || null;
    const bathrooms = $box.find("span").filter((_, el) => /bath/.test($(el).text())).first().text() || null;
    const bedrooms = $box.find("span").filter((_, el) => /bed/.test($(el).text())).first().text() || null;
    const livingrooms = $box.find("span").filter((_, el) => /reception/.test($(el).text())).first().text() || null;
    const image = $box.find("picture source").first().attr("srcset") || null;
    const agency = $("img[src*='agent']").first().attr("alt") ?? null;
    data.push({
      price: price ? parseInt(price.split(" ")[0].replace(/£|,/g, ""), 10) : null,
      priceCurrency: "Sterling pound £",
      url: url ? "https://www.zoopla.co.uk" + url.split("?")[0] : null,
      image: image ? image.split(":p")[0] : null,
      address: $box.find("address").first().text() || null,
      squareFt: sqFtText ? parseInt(sqFtText.split(" ")[0].replace(/[~,]/g, ""), 10) : null,
      numBathrooms: bathrooms ? parseInt(bathrooms.split(" ")[0], 10) : null,
      numBedrooms: bedrooms ? parseInt(bedrooms.split(" ")[0], 10) : null,
      numLivingRoom: livingrooms ? parseInt(livingrooms.split(" ")[0], 10) : null,
      description: $box.find("a:has(address) p").first().text() || null,
      justAdded: $box.find("div").filter((_, el) => $(el).text() === "Just added").length > 0,
      agency,
    });
  });
  return { search_data: data, total_pages: totalPages };
}

// ---------------- scrape functions (mirror the upstream reference verbatim) ----------------

export async function scrapeProperties(urls) {
  const out = [];
  for (const url of urls) {
    try {
      const html = await fetchHtml(url, "section[aria-labelledby='local-area']", { autoScroll: true });
      out.push(parseProperty(html));
    } catch (e) {
      console.warn(`error scraping ${url}: ${e.message}`);
    }
  }
  return out;
}

export async function scrapeSearch(scrapeAllPages, locationSlug, maxScrapePages = 10, queryType = "for-sale") {
  const firstUrl = `https://www.zoopla.co.uk/${queryType}/property/${locationSlug}`;
  const firstHtml = await fetchHtml(firstUrl, "p[data-testid='total-results']", { autoScroll: true });
  const first = parseSearch(firstHtml);
  const searchData = [...first.search_data];
  const maxPages = first.total_pages;
  const totalToScrape = scrapeAllPages || maxScrapePages >= maxPages ? maxPages : maxScrapePages;
  for (let p = 2; p <= totalToScrape; p++) {
    try {
      const html = await fetchHtml(`${firstUrl}?pn=${p}`, "div[data-testid='regular-listings']");
      searchData.push(...parseSearch(html).search_data);
    } catch (e) {
      console.warn(`failed page ${p}: ${e.message}`);
    }
  }
  return searchData;
}
